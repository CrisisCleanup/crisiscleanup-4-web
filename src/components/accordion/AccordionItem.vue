<template>
  <div :class="classes">
    <button :class="buttonClasses" @click="toggleOpen">
      <slot name="name">
        {{ name }}
      </slot>
      <font-awesome-icon :icon="isOpen ? openIcon : closedIcon" />
    </button>
    <div v-show="isOpen" :class="bodyClasses">
      <slot></slot>
    </div>
  </div>
</template>

<script>
import { reactive, toRefs, computed } from 'vue';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';
import {
  faPlus,
  faMinus,
  faChevronDown,
  faChevronRight,
} from '@fortawesome/free-solid-svg-icons';

library.add(faPlus, faMinus, faChevronDown, faChevronRight);

export default {
  name: 'AccordionItem',
  components: {
    FontAwesomeIcon,
  },
  props: {
    name: {
      type: String,
      required: true,
    },
    iconStyle: {
      type: String,
      default: 'chevron', // Options: 'chevron', 'plus-minus'
    },
    classes: {
      type: String,
      default: 'border-b border-gray-200',
    },
    buttonClasses: {
      type: String,
      default:
        'w-full text-left bg-white py-2 px-4 hover:bg-gray-100 focus:outline-none flex justify-between items-center',
    },
    bodyClasses: {
      type: String,
      default: 'py-2 px-4',
    },
  },
  setup(props) {
    const state = reactive({
      isOpen: false,
    });

    const toggleOpen = () => {
      state.isOpen = !state.isOpen;
    };

    const openIcon = computed(() =>
      props.iconStyle === 'chevron' ? 'chevron-down' : 'minus',
    );
    const closedIcon = computed(() =>
      props.iconStyle === 'chevron' ? 'chevron-right' : 'plus',
    );

    return {
      ...toRefs(state),
      toggleOpen,
      openIcon,
      closedIcon,
    };
  },
};
</script>
