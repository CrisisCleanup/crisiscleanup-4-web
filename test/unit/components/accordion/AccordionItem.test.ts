import { mount } from '@vue/test-utils';
import { describe, expect, it } from 'vitest';
import AccordionItem from '@/components/accordion/AccordionItem.vue';

describe('AccordionItem', () => {
  it('should render with name prop', () => {
    const wrapper = mount(AccordionItem, {
      props: {
        name: 'Test Name',
      },
    });
    expect(wrapper.text()).toContain('Test Name');
  });

  it('should render with name slot', () => {
    const wrapper = mount(AccordionItem, {
      props: {
        name: 'Test Name',
      },
      slots: {
        name: '<div>Slot Name</div>',
      },
    });
    expect(wrapper.text()).toContain('Slot Name');
  });

  it.todo('should render default slot content', () => {
    const wrapper = mount(AccordionItem, {
      props: {
        name: 'Test Name',
      },
      slots: {
        default: '<div>Inner Content</div>',
      },
    });
    expect(wrapper.html()).not.toContain('Inner Content'); // Initially hidden
    wrapper.find('button').trigger('click');
    expect(wrapper.html()).toContain('Inner Content'); // Should be shown after click
  });

  it('should toggle isOpen state when button is clicked', async () => {
    const wrapper = mount(AccordionItem, {
      props: {
        name: 'Test Name',
      },
    });
    const button = wrapper.find('button');
    expect(wrapper.vm.isOpen).toBe(false);

    await button.trigger('click');
    expect(wrapper.vm.isOpen).toBe(true);

    await button.trigger('click');
    expect(wrapper.vm.isOpen).toBe(false);
  });

  it('should change icon based on isOpen state', async () => {
    const wrapper = mount(AccordionItem, {
      props: {
        name: 'Test Name',
        iconStyle: 'chevron',
      },
    });
    const button = wrapper.find('button');
    expect(wrapper.find('.svg-inline--fa').attributes('data-icon')).toBe(
      'chevron-right',
    );

    await button.trigger('click');
    expect(wrapper.find('.svg-inline--fa').attributes('data-icon')).toBe(
      'chevron-down',
    );

    await button.trigger('click');
    expect(wrapper.find('.svg-inline--fa').attributes('data-icon')).toBe(
      'chevron-right',
    );
  });

  it('matches snapshot when closed', () => {
    const wrapper = mount(AccordionItem, {
      props: {
        name: 'Test Name',
      },
      slots: {
        default: '<div>Inner Content</div>',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('matches snapshot when open', async () => {
    const wrapper = mount(AccordionItem, {
      props: {
        name: 'Test Name',
      },
      slots: {
        default: '<div>Inner Content</div>',
      },
    });
    await wrapper.find('button').trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
