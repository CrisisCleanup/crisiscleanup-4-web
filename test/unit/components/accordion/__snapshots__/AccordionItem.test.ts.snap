// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccordionItem > matches snapshot when closed 1`] = `
"<div class="border-b border-gray-200"><button class="w-full text-left bg-white py-2 px-4 hover:bg-gray-100 focus:outline-none flex justify-between items-center">
    <!--v-if-->Test Name<svg class="svg-inline--fa fa-chevron-right" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chevron-right" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512">
      <path class="" fill="currentColor" d="M310.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L242.7 256 73.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"></path>
    </svg>
  </button>
  <div class="py-2 px-4" style="display: none;">
    <div>Inner Content</div>
  </div>
</div>"
`;

exports[`AccordionItem > matches snapshot when open 1`] = `
"<div class="border-b border-gray-200"><button class="w-full text-left bg-white py-2 px-4 hover:bg-gray-100 focus:outline-none flex justify-between items-center">
    <!--v-if-->Test Name<svg class="svg-inline--fa fa-chevron-down" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chevron-down" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
      <path class="" fill="currentColor" d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"></path>
    </svg>
  </button>
  <div class="py-2 px-4">
    <div>Inner Content</div>
  </div>
</div>"
`;
